File: python/docsearch/adapters/text.py
"""Plain-text adapter: no structure, blank-line paragraph blocks."""

from __future__ import annotations

from pathlib import Path

from ..blocks import Block, Extraction


def extract(path: Path, progress: object = None) -> Extraction:
    raw = path.read_text(encoding="utf-8", errors="replace")
    blocks: list[Block] = []
    offset = 0
    for para in raw.split("\n\n"):
        body = para.strip()
        if body:
            blocks.append(Block(heading_path=[], locator={"offset": offset}, text=body))
        offset += len(para) + 2
    return Extraction(
        title=path.stem,
        format="text",
        blocks=blocks,
        diagnostics={"structure_source": "none (blank-line paragraphs)"},
    )

File: internal/adapter/text/text.go
// Package text reads a plain-text file, which has no structure: every run of
// text between blank lines is one block. Ported from
// python/docsearch/adapters/text.py.
package text

import (
	"path/filepath"
	"strings"
	"unicode/utf8"

	"github.com/bamsammich/docsearch/internal/domain"
	"github.com/bamsammich/docsearch/internal/pystr"
)

// paragraphBreak separates blocks. Offsets count it as two characters.
const paragraphBreak = "\n\n"

// Extract reads the plain-text file at path.
func Extract(path string) (*domain.Extraction, error) {
	raw, err := pystr.ReadFile(path)
	if err != nil {
		return nil, err
	}
	blocks := []domain.Block{}
	offset := 0
	for _, para := range strings.Split(raw, paragraphBreak) {
		if body := pystr.Strip(para); body != "" {
			blocks = append(blocks, domain.NewOffsetBlock(nil, offset, body))
		}
		offset += utf8.RuneCountInString(para) + len(paragraphBreak)
	}
	title := pystr.Stem(filepath.Base(path))
	return domain.NewExtraction(title, "text", domain.SourceBlankLines, blocks), nil
}
